### PR TITLE
#DVFD9-41 Add error message if the JSON data is invalid

### PR DESCRIPTION
**Pull request by WilsonWu2100**
Add error message to the Drupal Error Log if the uploaded JSON data is invalid

### CODE REVIEW

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php [hunk 1]
```diff
@@ -231,6 +232,10 @@ protected function fetchData() {
     $data = '{}';
 
     if ($response) {
+      $dvfHelpers = new DvfHelpers();
```
Comment (jez500) [inline dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: Should probably use dependency injection
eg https://github.com/govCMS/dvf/blob/2.x/dvf_ckan/src/Plugin/Visualisation/Source/CkanResource.php#L111

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php [hunk 2]
```diff
@@ -231,6 +232,10 @@ protected function fetchData() {
...
+      if(!$dvfHelpers->validateJson($response)) {
+        $this->logger->error($this->t('This is invalid JSON file'));
```
Comment (jez500) [inline dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: @WilsonWu2100 this doesn't really give that much info for debugging, consider adding something like 
```
$this->logger->error($this->t('Unable to parse this json file :uri', [':uri' => $uri]));
```
Also this is something we probably want to notify the user about too, so maybe add
```
$this->messenger()->addError($this->t('Invalid JSON file provided'));
```

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php (hunk 1 above)
Comment (thery-dh) [inline dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: @WilsonWu2100 I believe it would be better to move this to a property, instantiated within the JsonFile constructor, e.g:
`$this->dvfHelpers = new DvfHelpers();`

This way if we need to use it elsewhere within the class we can just do:
`$this->dvfHelpers->validateJson()`

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php (hunk 2 above)
Comment (thery-dh) [inline dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: I suggest rewording to `This JSON file is invalid`

**Review by thery-dh — CHANGES_REQUESTED** [review]
@WilsonWu2100 good work, a few suggestions :)

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php (hunk 1 above)
Comment (WilsonWu2100) [reply dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: Fixed

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php (hunk 2 above)
Comment (WilsonWu2100) [reply dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: Done

**Comment by jez500** [comment]
@WilsonWu2100 this didn't address my additional comments, see above

* Add url to log message
* Display message on frontend to user

**Comment by WilsonWu2100** [comment]
@jez500  I've made the changes, please have a review, thank you

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php [hunk 3]
```diff
@@ -231,6 +244,11 @@ protected function fetchData() {
     $data = '{}';
 
     if ($response) {
+      if(!$this->dvfHelpers->validateJson($response)) {
+        $this->messenger()->addError('This JSON file is invalid, ' . file_create_url($uri));
+        $this->logger->error($this->t('This JSON file is invalid, :url',
+          [':url' => file_create_url($uri)]));
```
Comment (jez500) [inline dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: @WilsonWu2100 this assumes it is a local file, `$uri` could be a remote url, change as per my comment
```
    [':url' => $uri]
```

**Review by thery-dh — APPROVED** [review]

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php (hunk 1 above)
Comment (WilsonWu2100) [reply dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: @jez500  This is the old code, which has been updated already.

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php (hunk 2 above)
Comment (WilsonWu2100) [reply dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: This is resolved.

File: dvf_json/src/Plugin/Visualisation/Source/JsonFile.php (hunk 3 above)
Comment (WilsonWu2100) [reply dvf_json/src/Plugin/Visualisation/Source/JsonFile.php]: This is resolved.

**Review by jez500 — APPROVED** [review]
